src/polycap-profile.c

```diff
@@ -0,0 +1,234 @@
+			polynomialfit(4, 3, pc_x, pc_y, coeff);
+
+			//calculate shape coordinates
+			for(i=0;i<=nmax;i++){
+				profile->z[i] = length/nmax*i; //z coordinates, from 0 to length
+				profile->cap[i] = (rad_int[1]-rad_int[0])/length*profile->z[i] + rad_int[0]; //single capillary shape always conical
+				profile->ext[i] = coeff[0]+coeff[1]*profile->z[i]+coeff[2]*profile->z[i]*profile->z[i];
+			}
+			break;
+
+		case 2: //ellipsoidal; side with largest radius has horizontal tangent, other side points towards focal_dist corresponding to smallest external radius
+			if(rad_ext[1] < rad_ext[0]){ //focussing alignment
+				slope = rad_ext[1] / focal_dist[1];
+				b = (-1.*(rad_ext[1]-rad_ext[0])*(rad_ext[1]-rad_ext[0])-slope*length*(rad_ext[1]-rad_ext[0])) / (slope*length+2.*(rad_ext[1]-rad_ext[0]));
+				k = rad_ext[0] - b;
+				a = sqrt((b*b*length)/(slope*(rad_ext[1]-k)));
+				for(i=0;i<=nmax;i++){
+					profile->z[i] = length/nmax*i; //z coordinates, from 0 to length
+					profile->cap[i] = (rad_int[1]-rad_int[0])/length*profile->z[i] + rad_int[0]; //single capillary shape always conical
+					profile->ext[i] = sqrt(b*b-(b*b*profile->z[i]*profile->z[i])/(a*a))+k;
+				}
+			} else { //confocal (collimating) alignment
+				slope = rad_ext[0] / focal_dist[0];
+				b = (-1.*(rad_ext[0]-rad_ext[1])*(rad_ext[0]-rad_ext[1])-slope*length*(rad_ext[0]-rad_ext[1])) / (slope*length+2.*(rad_ext[0]-rad_ext[1]));
+				k = rad_ext[1] - b;
+				a = sqrt((b*b*length)/(slope*(rad_ext[0]-k)));
+				for(i=0;i<=nmax;i++){
+					profile->z[i] = length/nmax*i; //z coordinates, from 0 to length
+					profile->cap[i] = (rad_int[1]-rad_int[0])/length*profile->z[i] + rad_int[0]; //single capillary shape always conical
+					profile->ext[i] = sqrt(b*b-(b*b*profile->z[nmax-i]*profile->z[nmax-i])/(a*a))+k;
+				}
+			}
+			break;
+
+		default:
+			break;
+
+	}
+
+	return profile;
+
+}
+//===========================================
+// get a new profile from Laszlo's ASCII files
+polycap_profile* polycap_profile_new_from_file(const char *single_cap_profile_file, const char *central_axis_file, const char *external_shape_file)
+{
+	polycap_profile *profile;
+	FILE *fptr;
+	int i, n_tmp;
+	double sx, sy;
+
+	//single capillary profile
+	fptr = fopen(single_cap_profile_file,"r");
+	if(fptr == NULL){
+		printf("Could not open %s\n",single_cap_profile_file);
+		exit(2);
+		}
+	fscanf(fptr,"%d",&n_tmp);
+	
+	//Make profile of sufficient memory size
+	profile = malloc(sizeof(polycap_profile));
+	if(profile == NULL){
+		printf("Could not allocate profile memory.\n");
+		exit(1);
+	}
+	profile->nmax = n_tmp;
+	profile->z = malloc(sizeof(double)*(profile->nmax+1));
+	if(profile->z == NULL){
+		printf("Could not allocate profile->z memory.\n");
+		exit(1);
+	}
+	profile->cap = malloc(sizeof(double)*(profile->nmax+1));
+	if(profile->cap == NULL){
+		printf("Could not allocate profile->cap memory.\n");
+		exit(1);
+	}
+	profile->ext = malloc(sizeof(double)*(profile->nmax+1));
```

Comment:
Na wat valgrinden blijkt dat dit de eerste bug is: je malloc't wel `profile->ext`, maar je steekt er geen waarden in, wat later een hele hoop miserie geeft.

Reply:
Dang, wat een brainfart. Blijkbaar daar iets te veel stukken code gekopieerd, en dan vergeten aanpassen :) zal morgen fiksen.

Khad trouwens bij mij ook valgrind gebruikt, maar dat gaf niet echt nuttige fouten ofzo. Je gebruikt ook --tool=memcheck -v? Of iets anders?

Reply:
Nu we een shared library hebben, is de `polycap` executable in `build/src` een bash script geworden dat de eigenlijke executable opstart die ergens in `build/src/.libs` zit. Om nu valgrind aan de praat te krijgen, moet je dit `polycap` script een beetje aanpassen.

Open dit bestand, en zoek naar een lijn die lijkt op:

````bash
exec "$progdir/$program" ${1+"$@"}
````

In mijn geval, waarschijnlijk bij u ook, zit deze lijn in een functie genaamd `func_exec_program_core`.

Verander deze lijn naar:

````bash
exec valgrind --track-origins=yes "$progdir/$program" ${1+"$@"}
````

De `--track-origins=yes` optie voor valgrind laat toe te zien waar variabelen met niet geinitializeerd geheugen gealloceerd werden, wat ik dikwijls handig vindt...

Jammer genoeg wordt dit script overschreven wanneer `polycap` gehercompileerd wordt, waardoor je dit dus vaak zal moeten aanpassen...